tests/core/remodel-gadget/task.yaml

```diff
@@ -12,15 +12,34 @@ prepare: |
     . "$TESTSLIB"/core-config.sh
     #shellcheck source=tests/lib/systemd.sh
     . "$TESTSLIB"/systemd.sh
-    systemctl stop snapd.service snapd.socket
 
-    clean_snapd_lib
-    prepare_core_model
-    prepare_test_account valid-for-testing
-    prepare_test_model valid-for-testing-pc
-
-    # kick first boot again
-    systemctl start snapd.service snapd.socket
+    if [ "$SPREAD_REBOOT" = 0 ]; then
+        REBOOT_NEEDED=false
+        if os.query is-core18 && snap list snapd | NOMATCH ' x1 '; then
+            REBOOT_NEEDED=true
+        fi
+
+        systemctl stop snapd.service snapd.socket
+
+        clean_snapd_lib
+        prepare_core_model
+        prepare_test_account valid-for-testing
+        prepare_test_model valid-for-testing-pc
+
+        # kick first boot again
+        systemctl start snapd.service snapd.socket
+
+        if [ "$REBOOT_NEEDED" = true ]; then
+            echo "reboot when the system is ready"
+            for _ in $(seq 10); do
+                if "$TESTSTOOLS"/journal-state match-log "Waiting for system reboot"; then
+                     break
+                fi
+                sleep 1
+            done
+            REBOOT
+        fi
```

Comment:
This seems to be useful as a library function. Maybe even a bit of the code above, too, but I'm not sure about that.

Reply:
@mardy I'll try to re-implement this 'for' by using retry tool. I think I can simplify this core.
Thanks for the review